Add Image constructor specialised for rendering to a texture

tigregalis opened the pull request:
# Objective

Fixes #7358
Redo of #7360

Ergonomics. There's a bunch of enigmatic boilerplate for constructing a texture for rendering to, which could be greatly simplified for the external user-facing API.

## Solution

- Take part of the render_to_target example and turn it into a new constructor for `Image`, with minimal changes beyond the `Default` implementation.
- Update the render_to_target example to use the new API.

Strictly speaking, there are two small differences between the constructor and the example:

1. The example sets the `size` when initially constructing the `Image`, then `resize`s, but `resize` sets the `size` anyway so we don't need to do this extra step.

2. The example sets `Image.texture_descriptor.format` to `TextureFormat::Bgra8UnormSrgb`, but the default impl sets this to `TextureFormat::Rgba8UnormSrgb` via `wgpu::TextureFormat::bevy_default()`. I don't know what sort of impact this has, but it works on my machine.

I've deliberately chosen to only include `width` and `height` as parameters, but maybe it makes sense for some of the other properties to be exposed as parameters.

---

## Changelog

### Added

Added `Image::new_target_texture` constructor for simpler creation of render target textures.

---

Notes:

- This is a re-do of https://github.com/bevyengine/bevy/pull/7360 - there's some relevant discussion on code style there.
- The docs for the method want to refer to `bevy_render::camera::Camera` and bevy_render::camera::RenderTarget::Image`. `bevy_image` used to be part of `bevy_render` and was split out in the past, and `bevy_image` doesn't depend on `bevy_render`. What's the recommendation here?

tigregalis commented:
@BenjaminBrienen I've "rebased" this

chompaa commented on crates/bevy_image/src/image.rs:
```diff
@@ -761,6 +761,26 @@ impl Image {
+    pub fn new_target_texture(width: u32, height: u32) -> Self {
+        let mut image = Self::default();
+
+        image.resize(Extent3d {
```
I noticed you discussed being more explicit and constructing `Image` directly in your previous PR. Personally, I prefer that approach over using `resize`. Was there a reason this wasn't changed?

tigregalis replied:
I've pushed a new commit that inlines the logic for default and resize and tidied it it up - it is definitely clearer overall.

To answer your question, it's mostly because there wasn't really a clear direction given on the style, and the style for the constructors are already inconsistent within this file:

- `Default::default`: struct literal
- `Image::new`: start with `Default::default`
- `Image::transparent`: struct literal
- `Image::new_fill`: start with `Default:;default`
- (this) `Image::new_target_texture` (original): start with `Default:;default`
- (this) `Image::new_target_texture` (new): struct literal

I think part of the awkwardness comes from the fact that `TextureDescriptor` (a foreign type) doesn't have a default, and it's not clear what a sensible default should be. Somewhat related to this, is this comment:
```rs
pub struct Image {
    // ..
    // TODO: this nesting makes accessing Image metadata verbose. Either flatten out descriptor or add accessors
    pub texture_descriptor: TextureDescriptor<Option<&'static str>, &'static [TextureFormat]>,
    // ..
}
```

chompaa commented on crates/bevy_image/src/image.rs:
```diff
@@ -761,6 +761,26 @@ impl Image {
         value
     }
 
+    /// Create a new zero-filled image with a given size, which can be rendered to.
+    /// This is primarily for use as a render target for a [`Camera`](`bevy_render::camera::Camera`).
+    /// See [`RenderTarget::Image`](`bevy_render::camera::RenderTarget::Image`).
```
This fixes those pesky doc links :)

```suggestion
  /// Create a new zero-filled image with a given size, which can be rendered to. This is primarily
  /// for use as a render target for a [`Camera`]. See [`RenderTarget::Image`].
  ///
  /// [`Camera`]: https://docs.rs/bevy/latest/bevy/render/camera/struct.Camera.html
  /// [`RenderTarget::Image`]: https://docs.rs/bevy/latest/bevy/render/camera/enum.RenderTarget.html#variant.Image
```

JMS55 commented:
Shouldn't hard code to default the texture format. HDR cameras need a different format.

IceSentry reviewed (approved):
Can you also update headless_renderer, gpu_readback, and compute_shader_game_of_life

I think those are all the examples that use this pattern

IceSentry commented:
> Shouldn't hard code to default the texture format. HDR cameras need a different format.

Yeah, this should either take a `hdr` parameter or just a texture format.

tigregalis commented:
In the discord there were a few comments:

1. `BevyDefault` which is defined within `bevy_image` and only has a single implementer (`TextureFormat`) should be removed and any calls to `TextureFormat::bevy_default()` should be replaced with (now) `bevy_image::TEXTURE_FORMAT_SDR`.
2. Perhaps this new constructor really wants to live in `bevy_render`, either as a free function or as an extension trait.
3. Perhaps `TEXTURE_FORMAT_SDR` and `TEXTURE_FORMAT_HDR` should really live in `bevy_render`, but the other constructors on `Image` also want to use `TEXTURE_FORMAT_SDR` (as a "default" `TextureFormat`).

For now I have left the constructor in `bevy_image`, added a `hdr: bool` parameter, created the two constants, and have `ViewTarget::TEXTURE_FORMAT_HDR` point to `bevy_image::TEXTURE_FORMAT_HDR`.

tigregalis commented:
> Can you also update headless_renderer, gpu_readback, and compute_shader_game_of_life
> 
> I think those are all the examples that use this pattern.
> 
> (Sorry, that wasn't an approve, just a comment)

I've done this but the discrepancies make me think maybe these should have their own helper constructors, possibly with guidance in the doc comment.

- `headless_renderer`: additional texture usage `COPY_SRC`
- `gpu_readback`: additional texture usage `COPY_SRC` + `STORAGE_BINDING`, different texture format `R32Uint`, different asset usage `RENDER_WORLD`
- `compute_shader_game_of_life`: different texture usage `COPY_DST` + `STORAGE_BINDING` + `TEXTURE_BINDING`, different texture format `R32Float`, different asset usage `RENDER_WORLD`

IceSentry reviewed (changes requested):
Hey, sorry for forgetting about this PR. After looking at the change again, I think I'm almost happy with everything, but I realize now that maybe we should just have the texture format be a direct parameter to the function instead of an HDR flag. That way it can easily be used in other contexts and people can inline the `if hdr` pattern if they need it and if they don't it's pretty easy to just provide the right texture format.

I think it's fine if more advanced use cases just modify the resulting image like you did in the examples. It's still nicer to have the constructor as a starting point for that boilerplate.

tigregalis commented:
@IceSentry 

I've updated `Image::new_target_texture` to take `TextureFormat` parameter instead of `hdr` flag.

I've improved the documentation.

The examples I've updated are:
- `compute_shader_game_of_life`
- `render_to_texture`
- `headless_renderer`

Since example `gpu_readback` now uses the `Image::new_uninit` constructor, I've opted to revert that change and leave that is is; `Image::new_target_texture` felt like a misuse for it anyway.

IceSentry commented:
This looks good, but could you make the changes to add the TEXTURE_FORMAT consts to bevy_image in a separate PR? I think it's a worthwhile change, but I'd prefer if it wasn't bundled with a feature PR.

IceSentry commented on crates/bevy_image/src/image.rs:
```diff
@@ -817,6 +820,51 @@ impl Image {
+    ///
+    /// [`Camera`]: https://docs.rs/bevy/latest/bevy/render/camera/struct.Camera.html
+    /// [`RenderTarget::Image`]: https://docs.rs/bevy/latest/bevy/render/camera/enum.RenderTarget.html#variant.Image
+    pub fn new_target_texture(width: u32, height: u32, format: TextureFormat) -> Self {
```
The doc should probably add a note about the `RenderAssetUsages` being default. For some edge cases people might prefer using RENDER_WORLD instead of default. I think for simplicity's sake it's fine to not make it an argument, but documenting it so advanced users know they'll need to change it would be a good thing.

tigregalis replied:
Are you sure? In the primary `render_to_texture` example, which doesn't seem to be accessing the image from the CPU after creation:

With `RenderAssetUsages::default()` (`RenderAssetUsages::MAIN_WORLD | RenderAssetUsages::RENDER_WORLD`):

![image](https://github.com/user-attachments/assets/233da1ae-81a6-4d47-ba57-edbfbe87cc2d)

But with only `RenderAssetUsages::RENDER_WORLD` the first pass image isn't being lit properly?

![image](https://github.com/user-attachments/assets/666e8821-f3a2-4a0e-959a-b79b9486a08d)

tigregalis replied:
Could it be a lighting issue when something is `RENDER_WORLD` only? A similar problem occurs with example `headless_renderer`. Below, `cpu_image` is set to both `MAIN_WORLD | RENDER_WORLD`.

```rs
    // This is the texture that will be rendered to.
    let mut render_target_image =
        Image::new_target_texture(size.width, size.height, TextureFormat::bevy_default());
    render_target_image.texture_descriptor.usage |= TextureUsages::COPY_SRC;
    // render_target_image.asset_usage = Default::default(); // necessary so that the scene is lit
    let render_target_image_handle = images.add(render_target_image);

    // This is the texture that will be copied to.
    let mut cpu_image =
        Image::new_target_texture(size.width, size.height, TextureFormat::bevy_default());
    cpu_image.asset_usage = Default::default(); // necessary because it will crash otherwise
    let cpu_image_handle = images.add(cpu_image);
```

`render_target_image.asset_usage = RENDER_WORLD`:
![000](https://github.com/user-attachments/assets/5ef50a86-c44b-49ac-92f6-8e7da47de4c8)

`render_target_image.asset_usage = MAIN_WORLD | RENDER_WORLD`:
![old-000](https://github.com/user-attachments/assets/12f7fbf8-1992-4234-ae57-9e388d7607e8)

IceSentry replied:
Uh, that seems like a bug to me, I don't see why this would happen because of the usages. I guess let's ignore it for this PR for now.

tigregalis replied:
Alright, for this PR I've used `RenderAssetUsages::default()` as the default and documented it, since it makes the example code read "as intended". When that (lighting?) bug is fixed we can switch it to `RENDER_WORLD` only as the default and update the docs.

IceSentry commented:
Approved, but I would still like the texture format changes in bevy_image to be in a separate PR

tigregalis commented:
> Approved, but I would still like the texture format changes in bevy_image to be in a separate PR

Dropped the constants out of this PR.

tigregalis commented:
@janhohenheim Rebased and passing CI.
